[docs] Add ROM overview specification (#17633)
lowRISC/opentitan: sw/device/silicon_creator/rom/doc/rom_overview_specification.md

Reviewer: rendered md files should have [one sentence per line](https://opentitan.org/book/doc/contributing/style_guides/markdown_usage_style.html?highlight=markdown#general-markdown-style)

# Mask ROM Specification

## Overview

This document is a high-level overview of the features and responsibilities of the Mask ROM.

This document is not a comprehensive specification.
When appropriate, it refers to more complete specifications for various mask ROM responsibilities.
These specifications include:

> *   [Signature Verification Module [sigverify]](#sigverify)
> *   [Boot Policy Module [bootpolicy]](#bootpolicy)
> *   [Address space unlock module [unlock]](#unlock)
> *   [Bootstrap module [bootstrap]](#bootstrap)
> *   [Bootstrap decoding](#strapping)
> *   [Shutdown Module [shutdown]](#shutdown)
> *   [Chip-Specific Startup [startup]](#startup)
> *   [Mask ROM Pin Configuration [pinconfig]](#pinconfig)

Reviewer 2: My preference would be to have empty links for pages or sections that don't exist yet. It took me a minute to work out these links weren't just broken but intended to point to future content.

I'd link to have a link checker in CI at some point that is fragment aware. If I were to set this up, empty links would not cause a failure but broken fragments would.

```suggestion
*   [Signature Verification Module [sigverify]]()
*   [Boot Policy Module [bootpolicy]]()
*   [Address space unlock module [unlock]]()
*   [Bootstrap module [bootstrap]]()
*   [Bootstrap decoding]()
*   [Shutdown Module [shutdown]]()
*   [Chip-Specific Startup [startup]]()
*   [Mask ROM Pin Configuration [pinconfig]]()
```

Author: Done.

# Mask ROM Specification

> - Authors: cfrantz
> - Reviewers: alphan, moidx
> - Date: 2021-06-28
> - Status: Implmented

Reviewer: ```suggestion
```

## Overview

This document is a high-level overview of the features and responsibilities of the Mask ROM.

This document is not a comprehensive specification.  When appropriate, it refers to more complete specifications for various mask ROM responsibilities.  These specifications include:

> *   [Signature Verification Module [sigverify]](#sigverify)

Reviewer: nit: should this list use regular parentheses instead of square brackets for short names?

Author: I'd rather use square brackets.  Parentheticals are for providing additional information or side-commentary.

The square brackets are used because the words are meant to be read and to signal a reference to an external spec.

*   [Boot Policy Module [bootpolicy]](#bootpolicy)
*   [Address space unlock module [unlock]](#unlock)
*   [Bootstrap module [bootstrap]](#bootstrap)
*   [Bootstrap decoding](#strapping)
*   [Shutdown Module [shutdown]](#shutdown)
*   [Chip-Specific Startup [startup]](#startup)
*   [Mask ROM Pin Configuration [pinconfig]](#pinconfig)
*   [Memory Protection [memprotect]](memory_protection.md)

## Threat Model

## Dependencies

> The mask ROM is almost fully self-contained. All of the code in the mask ROM is located in //sw/device/silicon\_creator with a few exceptions.

Reviewer: nit
```suggestion
The mask ROM is almost fully self-contained. All of the code in the mask ROM is located in `//sw/device/silicon_creator` with a few exceptions.
```

We permit the following exceptions:

*   Base utility functions from `//sw/device/lib/base`.
*   The top\_level memory layout from `//hw/top_earlgrey/sw/autogen`.
*   Autogenerated register definitions (ie: `alert_handler_regs.h`, `flash_ctrl_regs.h`, etc).

> TODO: Once we convert the build system to bazel, we can use bazel’s visibility rules to enforce the dependency restrictions in the build.

Reviewer: ```suggestion
```

## Manufacturing Bring-Up

During bring-up while the device is in one of the TEST\_UNLOCKED&lt;n> states, certain countermeasures will not be enabled to facilitate testing and bring-up of the chip.

Notably:

*   The alert handler will not be configured or enabled in TEST\_UNLOCKED&lt;n>.

> *   TODO: [flash scrambling configuration]

Reviewer: nit: do we need this TODO? Is this already captured in an issue if it needs to be addressed?

## Keys

### Silicon Creator Keys

The Silicon Creator keys are provisioned when the silicon creator configures and initializes the signing environment.  This process is currently being designed and the details are beyond the scope of this document.  For now, the publicly available development FPGA keys are used as stand-ins for the silicon creator keys.

#### ROM.KEYS-SILICON-CREATOR-PROD

> These keys are RSA-3072 public keys with exponent 3 or F4.  The PROD keys are valid in all lifecycle states.

Reviewer: ```suggestion
These keys are RSA-3072 public keys with exponent 65537 (F4).
The PROD keys are valid in all lifecycle states.
```

Also applies to the remaining key types.

#### ROM.KEYS-SILICON-CREATOR-DEV

These keys are RSA-3072 public keys with exponent 3 or F4.  The DEV keys are valid in lifecycle states DEV and RMA.

#### ROM.KEYS-SILICON-CREATOR-BRINGUP

> These keys are RSA-3072 public keys with exponent 3 or F4.  The BRINGUP keys are valid in lifecycle states TEST\_UNLOCKED&lt;n> and RMA.

Reviewer: Do you want to update the code to use the name "bringup" or update this doc to use "test" for the sake of consistency?

Author: Changed to `TEST`.

> TBD: OTP - hash of valid key if we strike all others?

Reviewer: ```suggestion
```

## Boot Flow

The following sequence of events is executed in order.  ID labels are assigned to each step for reference.

### ROM.ASM-RESET-START

*   Clear all machine defined interrupts.
*   Zero all cpu registers.

See the [[startup]](#startup) specification for details.

### ROM.ASM-AST-INIT

> *   Copy the AST values from the CREATOR\_SW\_CFG\_AST\_CFG OTP words to the AST.

Reviewer: ```suggestion
*   Copy the AST values from the `CREATOR_SW_CFG_AST_CFG` OTP words to the AST.
```

See the [[startup]](#startup) specification for details.

### ROM.ASM-SRAM-INIT

*   Turn on the minimum level of entropy required to initialize memory scrambling.
*   Enable SRAM scrambling and initialize SRAM.

See the [[startup]](#startup) specification for details.

### ROM.ASM-PMP-INIT

*   Initialize the PMP.  Basic configuration:
    *   R+X for ROM regions.
    *   R for FLASH
    *   RW for RAM

See the [[memory protection]](memory_protection.md) specification for details.

### ROM.ASM-CRT-INIT

*   Zero .bss and copy .data to RAM.
*   Initialize stack pointer and global pointer.
*   Load the interrupt vector into mtvec.

> *   Jump to `mask\_rom\_main`.

Reviewer: ```suggestion
*   Jump to `rom_main`.
```

See the [[startup]](#startup) specification for details.

### ROM.WATCHDOG-INIT

*   Initialize the watchdog timer to instigate a reset after $timeout.  (should we read this value from OTP?)

### ROM.RST-REASON-INIT

*   Read the reset reason.  If power-on reset, init the retention RAM.

### ROM.SEC-MMIO-INIT

> *   Initialize the sec\_mmio library.  Rescan any OTPs read prior to this point to load the expectations into sec\_mmio’s register tracking.

Reviewer: ```suggestion
*   Initialize the `sec_mmio` library.  Rescan any OTPs read prior to this point to load the expectations into `sec_mmio`’s register tracking.
```

### ROM.PMP-INIT

*   Initialize the PMP module with values from ROM.ASM-PMP-INIT.

See the [[memory protection]](memory_protection.md) specification for details.

### ROM.SHUTDOWN-INIT

*   Initialize the shutdown module with values from OTP.
*   Loads alert configurations and alert class configurations from OTP into the alert handler block.

See the [[shutdown]](#shutdown) specification for details.

### ROM.PINMUX-INIT

*   Initialize pinmux to allow uart output and reading of the strap pins.

See the [[pin configuration]](#pinconfig) specification for details.

### ROM.UART-INIT

*   Initialize the uart to 115200 bps.

> *   Print the OpenTitan version identifier (e.g. git version of the mask rom code).
> *   Print the lifecycle state.

Reviewer: ```suggestion
```

(we print the BFV and LCV during shutdown)

### ROM.FLASH-INIT

*   Flash initializes its own scrambling parameters in hardware.  Make sure the init has finished by waiting for FLASH\_CTRL.STATUS.INIT\_WIP to be zero.

### ROM.BOOTSTRAP-MODE

*   Read the value of the strapping pins.
    *   If they indicate bootstrap mode and if bootstrap is permitted (via OTP ROM\_BOOTSTRAP\_EN), perform the bootstrap protocol as specified in the [Bootstrap Spec].
    *   For all other values, continue booting without bootstrap mode.

See the [[bootstrap]](#bootstrap) specification for details.

### ROM.KEYMGR-INIT

*   Initialize the key manager.

### ROM.BOOT-POLICY-READ

*   Read the boot policy from the flash INFO page and retention RAM and determine the parameters which will control booting the ROM\_EXT (which side to boot, which side is fallback, temporary side override and any anti-rollback parameters.

> See the [[bootpolicy]](#bootpolicy) specification for details.

Reviewer: We will probably use this logic in ROM_EXT. ROM prioritizes the slot with the newest (in terms of the security version) ROM_EXT.

Author: What is the current behavior here?  Read the boot policy and make sure the selected ROM_EXT meets the minimum version requirement from the policy?

### ROM.SIG-VERIFY

*   Using the boot policy, find and examine the ROM\_EXT manifest header.

>     *   Verify the correctness of the magic number ‘OTRE’.

Reviewer: ```suggestion
    *   Verify the correctness of the magic number ("OTRE" in ASCII).
```

    *   Verify the signature over the ROM\_EXT as specified in the [Sig-Verify Spec].

>         *   Verify header fields post signature verification: code boundaries, code start address, anti-rollback parameters, etc.

Reviewer: ```suggestion
        *   Verify header fields before signature verification: code boundaries, code start address, anti-rollback parameters, etc.
```

>         *   If validation fails, exit `main` with kErroMaskRomBootFailed, thus entering ROM.SHUTDOWN (also print the failed hash?)

Reviewer: ```suggestion
        *   If validation fails, exit `main` with kErroMaskRomBootFailed, thus entering ROM.SHUTDOWN.
```

        *   If validation succeeds proceed to the next step.

See the [[sigverify]](#sigverify) specification for details.

### ROM.BOOT-ROM\_EXT

*   Diversify the key manager with the software binding tag from the validated manifest.

> *   Print relevant ROM\_EXT boot information: version and hash?

Reviewer: ```suggestion
```

> *   Unlock flash execution in the PMP with the execution bounds from the manifest.

Reviewer: ```suggestion
*   Unlock flash execution in the flash controller and in the PMP with the execution bounds from the manifest.
```

> *   Verify sec\_mmio expectations and PMP configuration.

Reviewer: ```suggestion
*   Verify `sec_mmio` expectations and PMP configuration.
```

*   Jump to the ROM\_EXT.

### ROM.SHUTDOWN

*   Shutdown the chip according to the [Shutdown Spec].
*   Redact the fault code according to the OTP-defined redaction policy.
*   Print the redacted fault code to the UART.
*   Initiate a shutdown via the alert handler (SW triggerable alert).
*   Perform a best-effort shutdown in the event the alert handler fails to shutdown
    *   Disable (virtual-scrap) the crypto blocks (ie: keymanager).
    *   Disable access to flash.

>     *   Reconfigure PMP to disable access to everything but ROM.

Reviewer: ```suggestion
    *   Reconfigure PMP to disable access to everything but the shutdown infinite loop.
```
